src/segment_reshape/map_tool/segment_reshape_tool.py

```diff
@@ -0,0 +1,186 @@
+        self, location: QgsPointXY, mouse_button: Qt.MouseButton
+    ) -> None:
+        if mouse_button == Qt.LeftButton and self.pick_location_mode is True:
+            common_segment, layer = self._find_common_segment(location)
+
+            # No active layer or active layer feature found
+            if layer is None:
+                return
+
+            # No common segment found
+            if common_segment is None:
+                MsgBar.info(
+                    tr("No common segment found at location"),
+                    tr(
+                        "Features are not topologically connected "
+                        "or a single vertex was clicked"
+                    ),
+                )
+            else:
+                MsgBar.info(
+                    tr("Common segment found, changing to reshape mode"),
+                    success=True,
+                )
+                start_point = common_segment.startPoint()
+                self._change_to_reshape_mode()
+                self.old_segment_rubber_band.setToGeometry(
+                    QgsGeometry(common_segment), layer
+                )
+                self.temporary_new_segment_rubber_band.addPoint(
+                    QgsPointXY(start_point), True
+                )
+
+        elif mouse_button == Qt.LeftButton and self.reshape_mode is True:
+            self.new_segment_rubber_band.addPoint(location, True)
+            self.temporary_new_segment_rubber_band.reset()
+            self.temporary_new_segment_rubber_band.addPoint(location, True)
+
+        elif mouse_button == Qt.RightButton and self.reshape_mode is True:
+            new_geometry = self.new_segment_rubber_band.asGeometry()
+
+            # Reshape cancelled
+            if new_geometry.isEmpty():
+                self._change_to_pick_location_mode()
+                return
+
+            _, common_parts, edges = self.find_segment_results
+
+            reshape.make_reshape_edits(
+                common_parts, edges, QgsLineString(list(new_geometry.vertices()))
+            )
+
+            self._change_to_pick_location_mode()
+            self.canvas().refresh()
+
+            MsgBar.info(
+                tr("Features reshaped"),
+                success=True,
+            )
+
+    def _handle_mouse_move_event(self, location: QgsPointXY) -> None:
+        if self.reshape_mode is True:
+            self.temporary_new_segment_rubber_band.removeLastPoint(0, True)
+            self.temporary_new_segment_rubber_band.addPoint(location, True)
+
+    def _find_common_segment(
+        self, location: QgsPointXY
+    ) -> Tuple[Optional[QgsLineString], Optional[QgsVectorLayer]]:
+        LOGGER.info("Calculating common segment")
+
+        active_layer = iface.activeLayer()
+        if active_layer is None:
+            MsgBar.warning(tr("No active layer found"), tr("Activate a layer first"))
+            return None, None
+
+        identify_results = QgsMapToolIdentify(self.canvas()).identify(
```

Comment:
this might be problematic since it creates a new instance every time and never cleans it up (canvas takes ownership), maybe add the sub-tool as a private property in constructor?

Reply:
changed